File: gherkin-utils/javascript/src/GherkinDocumentWalker.ts
import { messages } from '@cucumber/messages'

export interface IFilters {
  acceptScenario?: (
    scenario: messages.GherkinDocument.Feature.IScenario
  ) => boolean
  acceptStep?: (step: messages.GherkinDocument.Feature.IStep) => boolean
  acceptBackground?: (
    background: messages.GherkinDocument.Feature.IBackground
  ) => boolean
  acceptRule?: (
    rule: messages.GherkinDocument.Feature.FeatureChild.IRule
  ) => boolean
  acceptFeature?: (feature: messages.GherkinDocument.IFeature) => boolean
}

export interface IHandlers {
  handleStep?: (step: messages.GherkinDocument.Feature.IStep) => void
  handleScenario?: (
    scenario: messages.GherkinDocument.Feature.IScenario
  ) => void
  handleBackground?: (
    background: messages.GherkinDocument.Feature.IBackground
  ) => void
  handleRule?: (
    rule: messages.GherkinDocument.Feature.FeatureChild.IRule
  ) => void
  handleFeature?: (feature: messages.GherkinDocument.IFeature) => void
}

const defaultFilters: IFilters = {
  acceptScenario: () => true,
  acceptStep: () => true,
  acceptBackground: () => true,
  acceptRule: () => true,
  acceptFeature: () => true,
}

export const rejectAllFilters: IFilters = {
  acceptScenario: () => false,
  acceptStep: () => false,
  acceptBackground: () => false,
  acceptRule: () => false,
  acceptFeature: () => false,
}

const defaultHandlers: IHandlers = {
  handleStep: () => null,
  handleScenario: () => null,
  handleBackground: () => null,
  handleRule: () => null,
  handleFeature: () => null,
}

export default class GherkinDocumentWalker {
  private readonly filters: IFilters
  private readonly handlers: IHandlers

  constructor(filters?: IFilters, handlers?: IHandlers) {
    this.filters = { ...defaultFilters, ...filters }
    this.handlers = { ...defaultHandlers, ...handlers }
  }

  public walkGherkinDocument(
    gherkinDocument: messages.IGherkinDocument
  ): messages.IGherkinDocument {
    if (!gherkinDocument.feature) {
      return null
    }

    const feature = this.walkFeature(gherkinDocument.feature)

    if (!feature) {
      return null
    }

    return messages.GherkinDocument.create({
      feature: feature,
      comments: gherkinDocument.comments,
      uri: gherkinDocument.uri,
    })
  }

  protected walkFeature(
    feature: messages.GherkinDocument.IFeature
  ): messages.GherkinDocument.IFeature {
    const keptChildren = this.walkFeatureChildren(feature.children)

    this.handlers.handleFeature(feature)

    const backgroundKept = keptChildren.find((child) => child.background)

    if (this.filters.acceptFeature(feature) || backgroundKept) {
      return this.copyFeature(
        feature,
        feature.children.map((child) => {
          if (child.background) {
            return messages.GherkinDocument.Feature.FeatureChild.create({
              background: this.copyBackground(child.background),
            })
          }
          if (child.scenario) {
            return messages.GherkinDocument.Feature.FeatureChild.create({
              scenario: this.copyScenario(child.scenario),
            })
          }
          if (child.rule) {
            return messages.GherkinDocument.Feature.FeatureChild.create({
              rule: this.copyRule(child.rule, child.rule.children),
            })
          }
        })
      )
    }

    if (keptChildren.find((child) => child !== null)) {
      return this.copyFeature(feature, keptChildren)
    }
  }

  private copyFeature(
    feature: messages.GherkinDocument.IFeature,
    children: messages.GherkinDocument.Feature.IFeatureChild[]
  ): messages.GherkinDocument.IFeature {
    return messages.GherkinDocument.Feature.create({
      children: this.filterFeatureChildren(feature, children),
      location: feature.location,
      language: feature.language,
      keyword: feature.keyword,
      name: feature.name,
      description: feature.description ? feature.description : undefined,
      tags: this.copyTags(feature.tags),
    })
  }

  private copyTags(
    tags: ReadonlyArray<messages.GherkinDocument.Feature.ITag>
  ): messages.GherkinDocument.Feature.ITag[] {
    return tags.map((tag) =>
      messages.GherkinDocument.Feature.Tag.create({
        name: tag.name,
        id: tag.id,
        location: tag.location,
      })
    )
  }

  private filterFeatureChildren(
    feature: messages.GherkinDocument.IFeature,
    children: messages.GherkinDocument.Feature.IFeatureChild[]
  ): messages.GherkinDocument.Feature.IFeatureChild[] {
    const copyChildren: messages.GherkinDocument.Feature.IFeatureChild[] = []

    const scenariosKeptById = new Map(
      children
        .filter((child) => child.scenario)
        .map((child) => [child.scenario.id, child])
    )

    const ruleKeptById = new Map(
      children
        .filter((child) => child.rule)
        .map((child) => [child.rule.id, child])
    )

    for (const child of feature.children) {
      if (child.background) {
        copyChildren.push(
          messages.GherkinDocument.Feature.FeatureChild.create({
            background: this.copyBackground(child.background),
          })
        )
      }

      if (child.scenario) {
        const scenarioCopy = scenariosKeptById.get(child.scenario.id)
        if (scenarioCopy) {
          copyChildren.push(scenarioCopy)
        }
      }

      if (child.rule) {
        const ruleCopy = ruleKeptById.get(child.rule.id)
        if (ruleCopy) {
          copyChildren.push(ruleCopy)
        }
      }
    }
    return copyChildren
  }

  private walkFeatureChildren(
    children: messages.GherkinDocument.Feature.IFeatureChild[]
  ): messages.GherkinDocument.Feature.IFeatureChild[] {
    const childrenCopy: messages.GherkinDocument.Feature.IFeatureChild[] = []

    for (const child of children) {
      let backgroundCopy: messages.GherkinDocument.Feature.IBackground = null
      let scenarioCopy: messages.GherkinDocument.Feature.IScenario = null
      let ruleCopy: messages.GherkinDocument.Feature.FeatureChild.IRule = null

      if (child.background) {
        backgroundCopy = this.walkBackground(child.background)
      }
      if (child.scenario) {
        scenarioCopy = this.walkScenario(child.scenario)
      }
      if (child.rule) {
        ruleCopy = this.walkRule(child.rule)
      }

      if (backgroundCopy || scenarioCopy || ruleCopy) {
        childrenCopy.push(
          messages.GherkinDocument.Feature.FeatureChild.create({
            background: backgroundCopy,
            scenario: scenarioCopy,
            rule: ruleCopy,
          })
        )
      }
    }

    return childrenCopy
  }

  protected walkRule(
    rule: messages.GherkinDocument.Feature.FeatureChild.IRule
  ): messages.GherkinDocument.Feature.FeatureChild.IRule {
    const children = this.walkRuleChildren(rule.children)

    this.handlers.handleRule(rule)

    const backgroundKept = children.find(
      (child) => child !== null && child.background !== null
    )
    const scenariosKept = children.filter(
      (child) => child !== null && child.scenario !== null
    )

    if (this.filters.acceptRule(rule) || backgroundKept) {
      return this.copyRule(rule, rule.children)
    }
    if (scenariosKept.length > 0) {
      return this.copyRule(rule, scenariosKept)
    }
  }

  private copyRule(
    rule: messages.GherkinDocument.Feature.FeatureChild.IRule,
    children: messages.GherkinDocument.Feature.FeatureChild.IRuleChild[]
  ): messages.GherkinDocument.Feature.FeatureChild.IRule {
    return messages.GherkinDocument.Feature.FeatureChild.Rule.create({
      id: rule.id,
      name: rule.name,
      description: rule.description ? rule.description : undefined,
      location: rule.location,
      keyword: rule.keyword,
      children: this.filterRuleChildren(rule.children, children),
    })
  }

  private filterRuleChildren(
    children: messages.GherkinDocument.Feature.FeatureChild.IRuleChild[],
    childrenKept: messages.GherkinDocument.Feature.FeatureChild.IRuleChild[]
  ): messages.GherkinDocument.Feature.FeatureChild.IRuleChild[] {
    const childrenCopy: messages.GherkinDocument.Feature.FeatureChild.IRuleChild[] = []
    const scenariosKeptIds = childrenKept
      .filter((child) => child.scenario)
      .map((child) => child.scenario.id)

    for (const child of children) {
      if (child.background) {
        childrenCopy.push(
          messages.GherkinDocument.Feature.FeatureChild.RuleChild.create({
            background: this.copyBackground(child.background),
          })
        )
      }
      if (child.scenario && scenariosKeptIds.includes(child.scenario.id)) {
        childrenCopy.push(
          messages.GherkinDocument.Feature.FeatureChild.RuleChild.create({
            scenario: this.copyScenario(child.scenario),
          })
        )
      }
    }

    return childrenCopy
  }

  private walkRuleChildren(
    children: messages.GherkinDocument.Feature.FeatureChild.IRuleChild[]
  ): messages.GherkinDocument.Feature.FeatureChild.IRuleChild[] {
    const childrenCopy: messages.GherkinDocument.Feature.FeatureChild.IRuleChild[] = []

    for (const child of children) {
      if (child.background) {
        childrenCopy.push(
          messages.GherkinDocument.Feature.FeatureChild.RuleChild.create({
            background: this.walkBackground(child.background),
          })
        )
      }
      if (child.scenario) {
        childrenCopy.push(
          messages.GherkinDocument.Feature.FeatureChild.RuleChild.create({
            scenario: this.walkScenario(child.scenario),
          })
        )
      }
    }
    return childrenCopy
  }

  protected walkBackground(
    background: messages.GherkinDocument.Feature.IBackground
  ): messages.GherkinDocument.Feature.IBackground {
    const steps = this.walkAllSteps(background.steps)
    this.handlers.handleBackground(background)

    if (
      this.filters.acceptBackground(background) ||
      steps.find((step) => step !== null)
    ) {
      return this.copyBackground(background)
    }
  }

  private copyBackground(
    background: messages.GherkinDocument.Feature.IBackground
  ): messages.GherkinDocument.Feature.IBackground {
    return messages.GherkinDocument.Feature.Background.create({
      id: background.id,
      name: background.name,
      location: background.location,
      keyword: background.keyword,
      steps: background.steps.map((step) => this.copyStep(step)),
      description: background.description ? background.description : undefined,
    })
  }

  protected walkScenario(
    scenario: messages.GherkinDocument.Feature.IScenario
  ): messages.GherkinDocument.Feature.IScenario {
    const steps = this.walkAllSteps(scenario.steps)
    this.handlers.handleScenario(scenario)

    if (
      this.filters.acceptScenario(scenario) ||
      steps.find((step) => step !== null)
    ) {
      return this.copyScenario(scenario)
    }
  }

  private copyScenario(
    scenario: messages.GherkinDocument.Feature.IScenario
  ): messages.GherkinDocument.Feature.IScenario {
    return messages.GherkinDocument.Feature.Scenario.create({
      id: scenario.id,
      name: scenario.name,
      description: scenario.description ? scenario.description : undefined,
      location: scenario.location,
      keyword: scenario.keyword,
      examples: scenario.examples,
      steps: scenario.steps.map((step) => this.copyStep(step)),
      tags: this.copyTags(scenario.tags),
    })
  }

  protected walkAllSteps(
    steps: messages.GherkinDocument.Feature.IStep[]
  ): messages.GherkinDocument.Feature.IStep[] {
    return steps.map((step) => this.walkStep(step))
  }

  protected walkStep(
    step: messages.GherkinDocument.Feature.IStep
  ): messages.GherkinDocument.Feature.IStep {
    this.handlers.handleStep(step)
    if (!this.filters.acceptStep(step)) {
      return null
    }
    return this.copyStep(step)
  }

  private copyStep(
    step: messages.GherkinDocument.Feature.IStep
  ): messages.GherkinDocument.Feature.IStep {
    return messages.GherkinDocument.Feature.Step.create({
      id: step.id,
      keyword: step.keyword,
      location: step.location,
      text: step.text,
      dataTable: step.dataTable,
      docString: step.docString,
    })
  }
}

File: gherkin-utils/java/src/main/java/org/gherkin/util/GherkinDocumentWalker.java
package org.gherkin.util;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Collectors;

import io.cucumber.messages.types.Background;
import io.cucumber.messages.types.Feature;
import io.cucumber.messages.types.FeatureChild;
import io.cucumber.messages.types.GherkinDocument;
import io.cucumber.messages.types.Rule;
import io.cucumber.messages.types.RuleChild;
import io.cucumber.messages.types.Scenario;
import io.cucumber.messages.types.Step;
import io.cucumber.messages.types.Tag;

public class GherkinDocumentWalker {
    private final IFilters filters;
    private final IHandlers handlers;

    GherkinDocumentWalker(IFilters filters, IHandlers handlers) {
        this.filters = filters == null ? new DefaultFilters() : filters;
        this.handlers = handlers == null ? new DefaultHandlers() : handlers;
    }

    GherkinDocumentWalker() {
        this.filters = new DefaultFilters();
        this.handlers = new DefaultHandlers();
    }

    public GherkinDocument walkGherkinDocument(GherkinDocument gherkinDocument) {
        if (gherkinDocument.getFeature() == null) {
            return null;
        }

        Feature feature = walkFeature(gherkinDocument.getFeature());

        if (feature == null) {
            return null;
        }

        return new GherkinDocument(gherkinDocument.getUri(), gherkinDocument.getFeature(),
                gherkinDocument.getComments());

    }

    protected Feature walkFeature(Feature feature) {
        List<FeatureChild> keptChildren = walkFeatureChildren(feature.getChildren());

        handlers.handleFeature(feature);

        List<FeatureChild> backgroundKept = keptChildren.stream().filter(fc -> fc.getBackground() != null).collect(Collectors.toList());

        if (filters.acceptFeature(feature) || !backgroundKept.isEmpty()) {
            return copyFeature(
                    feature,
                    feature.getChildren().stream().map((child) -> {
                        if (child.getBackground() != null) {
                            new FeatureChild(null, copyBackground(child.getBackground()), null);
                        }
                        if (child.getScenario() != null) {
                            return new FeatureChild(null, null, copyScenario(child.getScenario()));
                        }
                        if (child.getRule() != null) {
                            return new FeatureChild(copyRule(child.getRule(), child.getRule().getChildren()), null, null);
                        }
                        return null;
                    }).collect(Collectors.toList()));
        }

        if (keptChildren.stream().anyMatch(Objects::nonNull)) {
            return copyFeature(feature, keptChildren);
        }
        return null;
    }

    private Feature copyFeature(
            Feature feature,
            List<FeatureChild> children
    ) {
        return new Feature(
                feature.getLocation(),
                copyTags(feature.getTags()),
                feature.getLanguage(),
                feature.getKeyword(),
                feature.getName(),
                feature.getDescription(),
                filterFeatureChildren(feature, children));
    }

    private List<Tag> copyTags(List<Tag> tags) {
        return tags.stream().map(tag -> new Tag(tag.getLocation(), tag.getName(), tag.getId())).collect(Collectors.toList());
    }

    private List<FeatureChild> filterFeatureChildren(
            Feature feature,
            List<FeatureChild> children) {
        List<FeatureChild> copyChildren = new ArrayList<>();
        Map<String, FeatureChild> scenariosKeptById = children.stream()
                .filter(child -> child.getScenario() != null)
                .collect(Collectors.toMap(fc -> fc.getScenario().getId(), Function.identity()));

        Map<String, FeatureChild> ruleKeptById = children.stream()
                .filter(child -> child.getRule() != null)
                .collect(Collectors.toMap(fc -> fc.getRule().getId(), Function.identity()));

        for (FeatureChild child : feature.getChildren()) {
            if (child.getBackground() != null) {
                copyChildren.add(new FeatureChild(null, child.getBackground(), null));
            }
            if (child.getScenario() != null) {
                FeatureChild scenarioCopy = scenariosKeptById.get(child.getScenario().getId());
                if (scenarioCopy != null) {
                    copyChildren.add(scenarioCopy);
                }
            }
            if (child.getRule() != null) {
                FeatureChild ruleCopy = ruleKeptById.get(child.getRule().getId());
                if (ruleCopy != null) {
                    copyChildren.add(ruleCopy);
                }
            }
        }
        return copyChildren;
    }

    private List<FeatureChild> walkFeatureChildren(List<FeatureChild> children) {
        List<FeatureChild> childrenCopy = new ArrayList<>();

        for (FeatureChild child : children) {
            Background backgroundCopy = null;
            Scenario scenarioCopy = null;
            Rule ruleCopy = null;

            if (child.getBackground() != null) {
                backgroundCopy = walkBackground(child.getBackground());
            }
            if (child.getScenario() != null) {
                scenarioCopy = walkScenario(child.getScenario());
            }
            if (child.getRule() != null) {
                ruleCopy = walkRule(child.getRule());
            }

            if (backgroundCopy != null || scenarioCopy != null || ruleCopy != null) {
                childrenCopy.add(new FeatureChild(ruleCopy,
                        backgroundCopy,
                        scenarioCopy));
            }
        }

        return childrenCopy;
    }

    protected Rule walkRule(Rule rule) {
       List<RuleChild> children = walkRuleChildren(rule.getChildren());

       this.handlers.handleRule(rule);

        boolean backgroundKept = children.stream().anyMatch(child -> child != null && child.getBackground() != null);
        List<RuleChild> scenariosKept = children.stream().filter(child -> child != null && child.getScenario() != null).collect(
                Collectors.toList());

        if (this.filters.acceptRule(rule) || backgroundKept) {
            return copyRule(rule, rule.getChildren());
        }
        if (!scenariosKept.isEmpty()) {
            return copyRule(rule, scenariosKept);
        }
        return null;
    }

    private Rule copyRule(Rule rule, List<RuleChild> children) {
        return new Rule(rule.getLocation(),
                copyTags(rule.getTags()),
                rule.getKeyword(),
                rule.getName(),
                rule.getDescription(),
                filterRuleChildren(rule.getChildren(), children),
                rule.getId());
    }

    private List<RuleChild> filterRuleChildren(
            List<RuleChild> children,
            List<RuleChild> childrenKept) {
    List<RuleChild> childrenCopy = new ArrayList<>();
    Set<String> scenariosKeptIds = childrenKept
            .stream().filter(child -> child.getScenario() != null)
            .map(child -> child.getScenario().getId())
            .collect(Collectors.toSet());

        for (RuleChild child : children) {
            if (child.getBackground() != null) {
                childrenCopy.add(new RuleChild(copyBackground(child.getBackground()), null));
            }
            if (child.getScenario() != null && scenariosKeptIds.contains(child.getScenario().getId())) {
                childrenCopy.add(new RuleChild(null, copyScenario(child.getScenario())));
            }
        }
        return childrenCopy;
    }

    private List<RuleChild> walkRuleChildren(List<RuleChild> children) {
       List<RuleChild> childrenCopy = new ArrayList<>();

        for (RuleChild child : children) {
            if (child.getBackground() != null) {
                childrenCopy.add(new RuleChild(walkBackground(child.getBackground()), null));
            }
            if (child.getScenario() != null) {
                childrenCopy.add(new RuleChild(null, walkScenario(child.getScenario())));
            }
        }
        return childrenCopy;
    }

    protected Background walkBackground(Background background) {
        List<Step> steps = walkAllSteps(background.getSteps());
        handlers.handleBackground(background);

        if (this.filters.acceptBackground(background) || steps.stream().anyMatch(Objects::nonNull)) {
            return copyBackground(background);
        }
        return null;
    }

    private Background copyBackground(Background background) {
        return new Background(background.getLocation(),
                background.getKeyword(),
                background.getName(),
                background.getDescription(),
                background.getSteps().stream().map(step -> copyStep(step)).collect(Collectors.toList()),
                background.getId());
    }

    protected Scenario walkScenario(Scenario scenario) {
        List<Step> steps = walkAllSteps(scenario.getSteps());
        handlers.handleScenario(scenario);

        if (this.filters.acceptScenario(scenario) || steps.stream().anyMatch(Objects::nonNull)) {
            return copyScenario(scenario);
        }
        return null;
    }

    private Scenario copyScenario(Scenario scenario) {
        return new Scenario(scenario.getLocation(),
                copyTags(scenario.getTags()),
                scenario.getKeyword(),
                scenario.getName(),
                scenario.getDescription(),
                scenario.getSteps().stream().map(this::copyStep).collect(Collectors.toList()),
                scenario.getExamples(),
                scenario.getId());
    }

    protected List<Step> walkAllSteps(List<Step> steps) {
        return steps.stream().map(step -> walkStep(step)).collect(Collectors.toList());
    }

    protected Step walkStep(Step step) {
        handlers.handleStep(step);
        if (!filters.acceptStep(step)) {
            return null;
        }
        return copyStep(step);
    }

    private Step copyStep(Step step) {
        return new Step(step.getLocation(), step.getKeyword(), step.getText(), step.getDocString(), step.getDataTable(), step.getId());
    }
}
